File: 295-FindMedianfromDataStream/295-FindMedianfromDataStream.py
# Last updated: 4/21/2026, 1:25:38 PM
1import heapq
2class MedianFinder:
3
4    def __init__(self):
5        self.min_pq = []
6        self.max_pq = []
7
8    def addNum(self, num: int) -> None:
9        heapq.heappush(self.max_pq, -1 * num)
10        if self.min_pq and -1 * self.max_pq[0] > self.min_pq[0]:
11            heapq.heappush(self.min_pq, -1 * heapq.heappop(self.max_pq))
12        
13        if len(self.max_pq) > len(self.min_pq) + 1:
14            heapq.heappush(self.min_pq, -1 * heapq.heappop(self.max_pq))
15        elif len(self.min_pq) > len(self.max_pq) + 1:
16            heapq.heappush(self.max_pq, -1 * heapq.heappop(self.min_pq))
17
18    def findMedian(self) -> float:
19        if len(self.min_pq) == len(self.max_pq):
20            return (self.min_pq[0] + -1 * self.max_pq[0]) / 2
21        elif (len(self.max_pq) > len(self.min_pq)):
22            return -1 * self.max_pq[0]
23        else:
24            return self.min_pq[0]
25
26
27# Your MedianFinder object will be instantiated and called as such:
28# obj = MedianFinder()
29# obj.addNum(num)
30# param_2 = obj.findMedian()

File: 295-FindMedianfromDataStream/295-FindMedianfromDataStream.cpp
// Last updated: 9/11/2026, 6:41:24 PM
1class MedianFinder {
2public:
3    priority_queue<int> maxHeap;
4    priority_queue<int, vector<int>, greater<int>> minHeap;
5    MedianFinder() {
6        
7    }
8    
9    void addNum(int num) {
10        maxHeap.push(num);
11        int tmp = maxHeap.top();
12        maxHeap.pop();
13        minHeap.push(tmp);
14        if (maxHeap.size() - minHeap.size() == 2 || maxHeap.size() - minHeap.size() == -2) {
15            tmp = minHeap.top();
16            minHeap.pop();
17            maxHeap.push(tmp);
18        }
19    }
20    
21    double findMedian() {
22        if (maxHeap.size() == minHeap.size()) {
23            return (maxHeap.top() + minHeap.top()) / 2.0;
24        } else if (maxHeap.size() > minHeap.size()) {
25            return maxHeap.top();
26        } else {
27            return minHeap.top();
28        }
29    }
30};
31
32/**
33 * Your MedianFinder object will be instantiated and called as such:
34 * MedianFinder* obj = new MedianFinder();
35 * obj->addNum(num);
36 * double param_2 = obj->findMedian();
37 */